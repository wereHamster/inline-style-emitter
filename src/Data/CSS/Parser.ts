import { fromMaybe, maybe } from "../Maybe";

import { SDVFontFamily, CSSFontFaceDeclarationEx } from "./Internal/SDV/SDVFontFamily";

import { CSSStyle, CSSDeclarations, CSSRuleEx, styleHash, keyframesHash,
    CSSStyleDeclarationEx, CSSElementStyle, SDVAnimationName,
    CSSKeyframe } from "./Internal/Types";



export const cssStyleRules = (style: CSSStyle): CSSRuleEx[] =>
    style instanceof CSSElementStyle ? style.cssRules : styleRules(style);


// Parse a 'CSSStyleDeclarationEx' object into a list of 'CSSRuleEx' objects.
export const cssElementStyle = (x: CSSStyleDeclarationEx): CSSElementStyle =>
    new CSSElementStyle(styleRules(x));

export const styleRules = (x: CSSStyleDeclarationEx): CSSRuleEx[] => {
    const rules = [];
    extractStyleRules(rules, x, [], []);
    return rules;
};


// -----------------------------------------------------------------------------
// extractStyleRules
//
// Extract style rules from the style object and insert them into the rules"
// list. This function is recursive.

function extractStyleRules
( rules: CSSRuleEx[]
, s: CSSStyleDeclarationEx
, conditions: string[]
, suffixes: string[]
): void {
    const cssDeclarations: CSSDeclarations = {};

    for (const k in s) {
        const v = s[k];

        if (k[0] === ":") {
            // Pseudo classes and pseudo elements.
            extractStyleRules(rules, v, conditions, concat(suffixes, [k]));
        } else if (k[0] === "@") {
            if (k.startsWith("@media")) {
                extractStyleRules(rules, v, concat(conditions, [k]), suffixes);
            } else {
                // Ignore other @-rules.
                console.warn(`extractStyleRules: Ignoring unknown @-rule "${k}"`);
            }
        } else {
            cssDeclarations[k] = processValue(rules, k, v);
        }
    }

    tellCSSStyleRule(rules, cssDeclarations, conditions, suffixes);
}


type ValueProcessor = (rules: CSSRuleEx[], v: any) => undefined | number | string | string[];
const valueProcessors = {
    fontFamily(rules: CSSRuleEx[], v: SDVFontFamily): string | string[] {
        if (typeof v === "string") {
            return v;
        } else if (Array.isArray(v)) {
            return v.map(x => typeof x === "string" ? x : tellCSSFontFaceRule(rules, x));
        } else {
            return fromMaybe("", tellCSSFontFaceRule(rules, v));
        }
    },
    animationName(rules: CSSRuleEx[], v: SDVAnimationName): string {
        if (typeof v === "string") {
            return v;
        } else {
            return tellCSSKeyframesRule(rules, v);
        }
    }
};

// XXX: Why is it allowed to return undefined? This will essentially remove
// the property from the decls. Do we want to allow that?
export const processValue = (rules: CSSRuleEx[], k: string, v: any): undefined | string | string[] =>
    fromMaybe(v, maybe(stringifyValue(k, v), f => f(rules, v), valueProcessors[k]));

const stringifyValue = (key: string, prop: number | string | string[]): number | string | string[] =>
  (typeof prop !== "number" || prop === 0 || unitlessNumbers.has(key))
    ? prop : `${prop}px`;



const concat = <T>(xs: T[], ys: T[]): T[] =>
    [].concat.call([], xs, ys);

const concatMap = <A, B>(f: (x: A) => B[], xs: A[]): B[] =>
    xs.reduce((acc, x) => concat(acc, f(x)), []);


// In some parts of the code we treat CSSRuleEx[] as a mutable reference into
// which you can write. Sortof like a writer monad. Use these functions to
// append rules to the list.

const tellCSSStyleRule = (rules: CSSRuleEx[], cssDeclarations: CSSDeclarations, conditions: string[], suffixes: string[]): void => {
    if (Object.keys(cssDeclarations).length === 0) {
        return;
    }

    rules.push({
        type: 1,
        get hash() {
            const value = styleHash(cssDeclarations, concat(conditions, suffixes));
            Object.defineProperty(this, "hash", { value });
            return value;
        },
        conditions,
        suffixes,
        cssDeclarations,
        get className() {
            const value = "s" + this.hash;
            Object.defineProperty(this, "className", { value });
            return value;
        },
    });
};

export const tellCSSFontFaceRule = (rules: CSSRuleEx[], s: CSSFontFaceDeclarationEx): string => {
    // ASSERT notEmpty(s)

    const hash = styleHash(s, []);
    const {fontFamily, cssDeclarations} = maybe(
        {fontFamily: `f-${hash}`, cssDeclarations: Object.assign({}, s, {fontFamily: `f-${hash}`})},
        fontFamily => ({fontFamily, cssDeclarations: s}),
        s.fontFamily);

    rules.push({ type: 5, hash, cssDeclarations });

    return fontFamily;
};

export const tellCSSKeyframesRule = (rules: CSSRuleEx[], keyframes: CSSKeyframe[]): string => {
    // ASSERT notEmpty(s)

    const hash = keyframesHash(keyframes);
    const animationName = `a-${hash}`;

    rules.push({ type: 7, hash, keyframes, animationName });

    return animationName;
};


// Copied from Aphrodite, which apparently copied it from React. Who did React copy it from?
const unitlessNumbers: Set<string> = (() => {
    const prefixes = ["Webkit", "ms", "Moz", "O"];
    const explode = (key: string): string[] =>
        [key, ...prefixes.map(x => x + key.charAt(0).toUpperCase() + key.substring(1))];

    return new Set(concatMap(explode, [
        "animationIterationCount",
        "borderImageOutset",
        "borderImageSlice",
        "borderImageWidth",
        "boxFlex",
        "boxFlexGroup",
        "boxOrdinalGroup",
        "columnCount",
        "flex",
        "flexGrow",
        "flexPositive",
        "flexShrink",
        "flexNegative",
        "flexOrder",
        "gridRow",
        "gridColumn",
        "fontWeight",
        "lineClamp",
        "lineHeight",
        "opacity",
        "order",
        "orphans",
        "tabSize",
        "widows",
        "zIndex",
        "zoom",

        // SVG-related properties
        "fillOpacity",
        "floodOpacity",
        "stopOpacity",
        "strokeDasharray",
        "strokeDashoffset",
        "strokeMiterlimit",
        "strokeOpacity",
        "strokeWidth",
    ]));
})();
