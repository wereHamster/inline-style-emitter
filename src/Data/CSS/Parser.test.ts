/* global describe, it, expect */

declare var describe: any, it: any, expect: any;

import { CSSStyleDeclarationEx, CSSFontFaceDeclarationEx,
    CSSKeyframe } from "./Internal/Types";
import { cssElementStyle, styleRules, processValue } from "./Parser";



function testSnapshots1(inputs, f) {
    describe("snapshots", () =>
        inputs.forEach(input =>
            it(JSON.stringify(input), () =>
                expect(f(input)).toMatchSnapshot())));
}


const fontFamily1: CSSFontFaceDeclarationEx = {
    fontFamily: "FontAwesome",
    src: "url(./FontAwesome.woff2)",
};
const fontFamily2: CSSFontFaceDeclarationEx = {
    src: ["url(./FontAwesome.woff)", "url(./FontAwesome.woff2)"],
    fontStyle: "italic",
};
const animationName1: CSSKeyframe[] = [[0, {}]];

describe("cssElementStyle", () => {
    testSnapshots1(
        [ {}
        , {color: "red"}
        , {color: "red", ":hover": {}}
        , {color: "red", ":hover": {color: "blue"}}
        , {"@media (max-width:1px)": {}}
        , {"@media (max-width:1px)": {color: "red"}}
        ], cssElementStyle);
});

describe("styleRules", () => {
    describe("number of generated rules", () => {
        const expectLength = (n: number, x: CSSStyleDeclarationEx): void =>
            expect(styleRules(x).length).toBe(n);

        it("should return an empty list if the style is empty", () =>
            expectLength(0, {}));

        it("should return a single rule when the object has only local style declarations", () =>
            expectLength(1, {color: "red"}));

        it("should ignore empty nested pseudo components", () =>
            expectLength(1, {color: "red", ":hover": {}}));

        it("should parse nested pseudo components", () =>
            expectLength(2, {color: "red", ":hover": {color: "blue"}}));

        it("should ignore media query conditional with empty content", () =>
            expectLength(0, {"@media (max-width:1px)": {}}));

        it("should parse media query conditional", () =>
            expectLength(1, {"@media (max-width:1px)": {color: "red"}}));

        it("should generate an additional rule for fontFamily (single object)", () =>
            expectLength(2, {fontFamily: fontFamily1, color: "green"}));

        it("should generate an additional rule for fontFamily when the base object is otherwise empty", () =>
            expectLength(2, {fontFamily: fontFamily1}));

        it("should generate additional rules for fontFamily (list of objects)", () =>
            expectLength(3, {fontFamily: [fontFamily1, fontFamily2]}));
    });


    testSnapshots1(
        [ {}
        , {color: "red"}
        , {color: "red", ":hover": {}}
        , {color: "red", ":hover": {color: "blue"}}
        , {"@media (max-width:1px)": {}}
        , {"@media (max-width:1px)": {color: "red"}}
        ], styleRules);
});


describe("valueProcessors", () => {
    describe("margin", () => {
        it("leaves zero untouched", () => {
            expect(processValue([], "margin", 0)).toBe(0);
        });
        it("leaves strings untouched", () => {
            expect(processValue([], "margin", "0 auto")).toBe("0 auto");
        });
        it("appends 'px' suffix to non-zero numbers", () => {
            expect(processValue([], "margin", 1)).toBe("1px");
        });
    });
    describe("fontFamily", () => {
        it("leaves strings untouched", () => {
            expect(processValue([], "fontFamily", "sans")).toBe("sans");
        });
        it("converts an object into a string", () => {
            expect(processValue([], "fontFamily", fontFamily1)).toBe("FontAwesome");
        });
        it("converts an mixed array into strings", () => {
            expect(processValue([], "fontFamily", ["sans", fontFamily1])).toEqual(["sans", "FontAwesome"]);
        });
    });
    describe("flex", () => {
        it("leaves non-zero numbers untouched", () => {
            expect(processValue([], "flex", 1)).toBe(1);
        });
    });
    describe("WebkitFlex", () => {
        it("leaves non-zero numbers untouched", () => {
            expect(processValue([], "WebkitFlex", 1)).toBe(1);
        });
    });
    describe("animationName", () => {
        it("leaves strings untouched", () => {
            expect(processValue([], "animationName", "fancy-animation-1")).toBe("fancy-animation-1");
        });
        it("converts an object into a string", () => {
            expect(processValue([], "animationName", animationName1)).toBe("a-2cb14fb710d5b162");
        });
    });
});
