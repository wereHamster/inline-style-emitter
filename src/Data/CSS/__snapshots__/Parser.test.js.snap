exports[`cssElementStyle snapshots {"@media (max-width:1px)":{"color":"red"}} 1`] = `
CSSElementStyle {
  "cssRules": Array [
    Object {
      "className": "s43660e9eeeba4977",
      "conditions": Array [
        "@media (max-width:1px)"
      ],
      "cssDeclarations": Object {
        "color": "red"
      },
      "hash": "43660e9eeeba4977",
      "suffixes": Array [],
      "type": 1
    }
  ]
}
`;

exports[`cssElementStyle snapshots {"@media (max-width:1px)":{}} 1`] = `
CSSElementStyle {
  "cssRules": Array []
}
`;

exports[`cssElementStyle snapshots {"color":"red",":hover":{"color":"blue"}} 1`] = `
CSSElementStyle {
  "cssRules": Array [
    Object {
      "className": "s46576037294a675a",
      "conditions": Array [],
      "cssDeclarations": Object {
        "color": "blue"
      },
      "hash": "46576037294a675a",
      "suffixes": Array [
        ":hover"
      ],
      "type": 1
    },
    Object {
      "className": "s29ec1a4d45b6d719",
      "conditions": Array [],
      "cssDeclarations": Object {
        "color": "red"
      },
      "hash": "29ec1a4d45b6d719",
      "suffixes": Array [],
      "type": 1
    }
  ]
}
`;

exports[`cssElementStyle snapshots {"color":"red",":hover":{}} 1`] = `
CSSElementStyle {
  "cssRules": Array [
    Object {
      "className": "s29ec1a4d45b6d719",
      "conditions": Array [],
      "cssDeclarations": Object {
        "color": "red"
      },
      "hash": "29ec1a4d45b6d719",
      "suffixes": Array [],
      "type": 1
    }
  ]
}
`;

exports[`cssElementStyle snapshots {"color":"red"} 1`] = `
CSSElementStyle {
  "cssRules": Array [
    Object {
      "className": "s29ec1a4d45b6d719",
      "conditions": Array [],
      "cssDeclarations": Object {
        "color": "red"
      },
      "hash": "29ec1a4d45b6d719",
      "suffixes": Array [],
      "type": 1
    }
  ]
}
`;

exports[`cssElementStyle snapshots {} 1`] = `
CSSElementStyle {
  "cssRules": Array []
}
`;

exports[`styleRules snapshots {"@media (max-width:1px)":{"color":"red"}} 1`] = `
Array [
  Object {
    "className": "s43660e9eeeba4977",
    "conditions": Array [
      "@media (max-width:1px)"
    ],
    "cssDeclarations": Object {
      "color": "red"
    },
    "hash": "43660e9eeeba4977",
    "suffixes": Array [],
    "type": 1
  }
]
`;

exports[`styleRules snapshots {"@media (max-width:1px)":{}} 1`] = `Array []`;

exports[`styleRules snapshots {"color":"red",":hover":{"color":"blue"}} 1`] = `
Array [
  Object {
    "className": "s46576037294a675a",
    "conditions": Array [],
    "cssDeclarations": Object {
      "color": "blue"
    },
    "hash": "46576037294a675a",
    "suffixes": Array [
      ":hover"
    ],
    "type": 1
  },
  Object {
    "className": "s29ec1a4d45b6d719",
    "conditions": Array [],
    "cssDeclarations": Object {
      "color": "red"
    },
    "hash": "29ec1a4d45b6d719",
    "suffixes": Array [],
    "type": 1
  }
]
`;

exports[`styleRules snapshots {"color":"red",":hover":{}} 1`] = `
Array [
  Object {
    "className": "s29ec1a4d45b6d719",
    "conditions": Array [],
    "cssDeclarations": Object {
      "color": "red"
    },
    "hash": "29ec1a4d45b6d719",
    "suffixes": Array [],
    "type": 1
  }
]
`;

exports[`styleRules snapshots {"color":"red"} 1`] = `
Array [
  Object {
    "className": "s29ec1a4d45b6d719",
    "conditions": Array [],
    "cssDeclarations": Object {
      "color": "red"
    },
    "hash": "29ec1a4d45b6d719",
    "suffixes": Array [],
    "type": 1
  }
]
`;

exports[`styleRules snapshots {} 1`] = `Array []`;
